[Scheduler] [Component] Scheduler

Guikingone opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Implement https://github.com/symfony/symfony/issues/36821
| License       | MIT
| Doc PR        | To define

Hi everyone 👋 

New year, new idea, first, let me explain why I open this PR as a draft, this new "experimentation" is mostly a draft as the framework integration is not completely done (I've some work pushed on a private repository). 

So, what is it? 

First, it's a new component, inspired by what Laravel can do with Illuminate/Scheduling but more important, it's inspired by modern platform that allows us to deploy our apps, in most platform, we can define tasks (cron and so on) and if we deploy our apps to "old" platform, we need to edit a crontab file where we need to declare all of our tasks before executing them (and waiting to fail?), the `Scheduler` component is here to solve most of the problems that can comes with scheduled tasks: 

- What if my platform does not support cron?
- What if I need to plan tasks using Http? 
- What if I need to store tasks in a queue?
- Etc...

So, what's inside? 

First, we've a `Scheduler`, that's the entrypoint when it comes to scheduling a `Task` (this one can be a Http one, Cron, command, queue and so on), once scheduled, a `Worker` handle a task and execute it. 

For now, this experimentation is not complete (the Worker need to be improved, the Messenger "bridge" also and so on) and this PR lack what I'm working on when it comes to the framework integration, this PR is opened in order to discuss and see if it could be a good idea to work on or if it's preferable to stop it now 🙂 

Thanks for the feedback 👋 

PS: I know that SymfonyCloud use YAML to declare tasks but we don't always use it and the `Scheduler` idea is here to give a "standard" solution. 

Here's a small example: 

**Standalone**

```php
<?php 

use ...; 

$scheduler = new Scheduler('Europe/Paris');

$tasks = new ShellTask('app.test', 'echo Symfony', '*/5 * * * *');
$scheduler->schedule($task);
```

**Framework**

```yaml
framework:
  scheduler:
    timezone: 'Europe/Paris'
    transports:
      local:
        dsn: 'local://'
    schedulers:
      foo:
        transport: 'local'
    tasks:
      app.foo:
        scheduler: 'foo'
        type: 'shell'
        expression: '* * * * *'
        options:
          command: 'echo Symfony!'
```

Roadmap
---------

| Features            | State
| ------------- | ---
| Being able to schedule Symfony commands using the Command FQCN  | ✅ 
| Adding shortcuts to create "default" cron expressions | ✅
| Add methods that can trigger some logic before/after the execution of the task | ✅
| Being able to export the scheduled tasks using txt and json formats | ✅
| Use transports to ease the configuration | ✅  
| Being able to create cron files  | ✅ 
| Schedule tasks for specific duration  | ✅
| Schedule tasks for specific execution date  | ➡️ 
| Adding task (and pool) priority (*Pools removed*) | ✅
| Track and optimise tasks execution  | ➡️ 
| Allow to trigger tasks using a specific path (like fragments)  | ➡️ 
| Scheduler factory | ✅
| Cloud platform bridges | ➡️ 
| FrameworkBundle integration | ➡️ 
| WebProfilerBundle integration | ➡️ 
| Notifier integration | ➡️
| Documentation | *TODO*

**Core features:**

***Transports (under active work since January)***

- [X] Local
- [X] Null
- [X] Google Cloud Scheduler
- [x] Redis
- [x] Doctrine (under active work test)
- [ ] Kubernetes (under active work)
- [ ] Nomad (under active work)

***Commands***

- `bin/console scheduler:consume foo, bar --limit 10 --time-limit 10`
- `bin/console scheduler:export foo --directory /srv/app --format json --filename foo --limit 10`
- `bin/console scheduler:generate foo, bar --directory var/schedulers/export`
- `bin/console scheduler:get-tasks foo --expression * * * * * --state paused`
- `bin/console scheduler:list-failed`
- `bin/console scheduler:reboot foo --retry 10`
- `bin/console scheduler:run foo --name app.foo --expression * * * * *`

***Kernel & classes as Scheduler "entrypoints" (Added on 18/05/2020)***

`Scheduler` defines a `SchedulerAwareInterface` which allow virtually any class to be a entrypoint to define new tasks: 

```php
<?php

/*
 * This file is part of the Symfony package.
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Symfony\Component\Scheduler;

/**
 * @author Guillaume Loulier <contact@guillaumeloulier.fr>
 */
interface SchedulerAwareInterface
{
    public function schedule(SchedulerRegistryInterface $registry): void;
}
```

This interface is auto configured and every "class" that implement in will receive the `SchedulerRegistry` which contain every declared scheduler. 

***Worker subscribers | Scheduler subscribers*** (Added on 21/05/2020)

`SchedulerSubscriber` are specific subscribers which are registered on every|specific scheduler(s), the registration is done thanks to `getSubscribedSchedulers()`. 

```php
final class FooSchedulerSubscriber implements SchedulerSubscriberInterface
{
    /**
     * {@inheritdoc}
     */
    public static function getSubscribedEvents()
    {
        return ...;
    }

    /**
     * {@inheritdoc}
     */
    public static function getSubscribedSchedulers(): array
    {
        return ['*']; // But can also be ['foo', 'bar']
    }

    // ...
}
```

Guikingone committed:
feat(Component): Scheduler

stephane-lou commented:
I actually made a private bundle that does this. It doesn't use Messenger (for now) but allows me to schedule symfony commands using a graphical interface.
Currently I made a main command (taskamanager:execute) that checks every minute if there are scheduler commands to execute.

I also think this would be a great new component !

Guikingone commented:
Hi @stephane-lou 👋 

For now, the Messenger implementation is mainly here in order to allow to send tasks ti queue like RabbitMQ and build something similar to what Notifier does, lately, I've planned to build something that allows us to do the following process: 

- Send a task to RabbitMQ queue
- Consume the message thanks to a handler when the task expression is due
- If the message task is successful, send a notification (think Slack and so on)

For the graphical interface, I think it could be a great idea (using Security of course 😉 ), this way, we can easily gave the access to this interface to a specific team department that allows them to create task without specific knowledge of the internal logic 🙂

kbond commented:
Wow, what timing, I just open sourced a bundle for this: https://github.com/kbond/schedule-bundle

Preparing for a 1.0 release this week.

I'd be happy to help contribute of this is something that is decided should be in core.

Guikingone commented:
Oups, wrong click, sorry for the ping FabPot :(

fabpot commented:
The ping is actually interesting... as I do have a cron component locally. It was supposed to be announced at a Symfony conference, but I decided to release the Notifier one first. Might come soon though.

monteiro commented:
Thanks a lot for contributing @Guikingone.

Besides handling crons, I think this could also be interesting for one time tasks. For example, after 3 hours on a specific action, send an e-mail to try to convert a the customer instead of using a CRON.

Guikingone commented:
@monteiro Actually, that's a planned feature, I've also planned to being able to create a pool of tasks (think about user specific tasks) that can be executed anytime 🙂

Guikingone commented:
Hi everyone 👋 

Here's a small list of recently added features (that need to be completely tested): 

- Task pools
- Task pools & tasks priority
- Failed tasks list (the `Worker` will retrieve it when a runner throw an `Error`).

Guikingone commented:
@monteiro Regarding your feedback, I've pushed a refactor that allow to pass a valid expression handled by `strtotime` or even a `\DatetimeInterface`, the actual CRON lib allow to calculate the due date depending on a valid expression or a `\DatetimeInterface`

jderusse commented on src/Symfony/Component/Scheduler/Messenger/TaskMessageHandler.php:
```diff
@@ -0,0 +1,46 @@
+    {
+        $task = $message->getTask();
+
+        if (!CronExpression::factory($task->get('expression'))->isDue(new \DateTimeImmutable())) {
```
What if the messages took 1 hours to be proceed? Maybe `isDue` should take the date of creation of the task instead of `now`

Guikingone replied:
Improved

jderusse commented on src/Symfony/Component/Scheduler/Messenger/TaskMessageHandler.php:
```diff
@@ -0,0 +1,46 @@
+    {
+        $task = $message->getTask();
+
+        if (!CronExpression::factory($task->get('expression'))->isDue(new \DateTimeImmutable())) {
```
Shouldn't you take timezone into account?

Guikingone replied:
Improved

jderusse commented on src/Symfony/Component/Scheduler/Scheduler.php:
```diff
@@ -0,0 +1,170 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function getPools(): array
```
This is not used... I'm not sure to undersatnd the purpose of task pools

Guikingone replied:
For now, the pools are a draft, the idea is to create a task "group" which allow to group the tasks linked to something specific, think about task which are linked to "user creation" for example, you may want to define 2-3 tasks: 

- post-registration email
- same as first but 24H later if the account is not activated
- Welcome emails when the account is validated

Using a pool can be useful as we can regroup every tasks using the id of the user for example and schedule it every time a new user is created.

Guikingone replied:
For clarity, I've removed the idea of pools, the `TaskList` is enough 🙂

jderusse reviewed (commented):
A common issue with scheduler in a cluster, is asserting that the task will be schedule. And be scheduled once.

IMHO the component should at least:
- store the last scheduled time in order to re-run missed schedules
- provide a mechanism to avoid scheduling the same task twice.

Guikingone commented:
@jderusse For the first point, that's actually the case, every time a task is runned, the subscriber define a new 'last execution date": 

```php
public function onTaskExecuted(TaskExecutedEvent $event): void
{
    $task = $event->getTask();

    $task->set('last_execution', new \DateTimeImmutable());
}
```

For the second point, that's an interesting point that I've not worked on directly, for now, a task cannot have the same name as an existing one, I'm gonna invest on this point 🙂

bigfoot90 commented:
For tasks that should be executed once per time, maybe an integration with the [Lock Component](https://symfony.com/doc/current/components/lock.html#the-owner-of-the-lock) can be interesting.
Locks can be defined as `*.pid` file (**locally per server node**) or in a shared database/service like `mysql`, `memcache` or `redis` (**globally per application**)

Guikingone commented:
@bigfoot90 I'm not sure that adding a new dependency (even if it's a Symfony one) its a good idea, I'm pretty sure we can find a solution without using Lock.

bigfoot90 commented:
It should not be a hard dependecy, but a soft dependency (mentioned in **suggested** inside `composer.json`)
Then a small `How to implement locking tasks` guide that says:
Run `composer require symfony/lock` implement `LockingTaskImplement` in your Task class.

Guikingone commented:
Hi everyone 👋 

First, I want to thanks everyone for the first feedbacks and @nicolas-grekas for the mention on the Sf meetup slides 😊 

I've pushed a refactoring which remove the notion of `TaskPool`, the `TaskList` already play this role, second, I've started an integration of the Lock component (as mentioned by @bigfoot90 and @jderusse), It's not fully in place but I'm gonna improve the tests and continue to test the integration.

Third point, I've push a first functional implementation of the `Exporter` using the Serializer (for formats like XML and JSON), I've planned to take a moment to look at a CLI "formatter" but I'm not sure if it worth the works. 

Fourth, what's the plan now? 

As I've a first functional component, I'm gonna polish the tests, finish every case that I can see by far, continue on the framework integration and improve the perfs and so on 🙂

Guikingone commented:
Hi everyone, 

I've just added a roadmap in the first post, the idea is to list the features that I've planned to add, which are done and so on.

Guikingone commented:
Ok, new day, new feature, I've just pushed an `ExpressionFactory` which is capable of building CRON expressions using shortcuts, as I don't know if the FabPot "CRON component" will support both `at` and `batch` syntax, I don't have shortcuts for it.

Guikingone commented:
Hi everyone 👋 

New day, new feature (again, I know 😄 ), I've focused on bringing a DataCollector, for now, it collect the data during the request using a `TraceableScheduler` (I've planned to do the same thing for `Worker`), I'm not sure about what to collect by far so I've built a simple array which contains the core data linked to a task. 

I've also worked on a core feature of the component, being able to execute tasks using a specific route (like fragments), I'm not sure about the implementation but definitively a fun moment playing with this. 

Last but not least, I've improved the output handling when a runner finish to run a task, by default, the runner can return a string or null, the output is also send to an event. 

Don't hesitate to give feedbacks 🙂

Guikingone commented:
Hi everyone 👋 

New day, new feature, I've focused the work on the DX, the Scheduler has now an interface which define core methods (resume, pause, etc). 

Most of the logic is tested and in place, I've started 3 bridges for the following platforms: 

- Google Cloud Platform
- Amazon Web Services
- Microsoft Azure

I've also planned to build a bridge for SymfonyCloud but I don't know if there's a public API/SDK which allows us to plan cron tasks, for now and if it's not possible, this bridge will stay as a "planned one", for the bridges, the Google one is prettymuch  in place, I've used the HttpClient component to call the REST API, the Google approach is pretty simple when it comes to scheduling tasks, Microsoft is a bit harder (and I don't talk about Amazon for now ... 😄 ). 

That's all for today, don't hesitate to give feedbacks 😉

Guikingone commented on src/Symfony/Component/Scheduler/Command/ExportTaskCommand.php:
```diff
@@ -0,0 +1,118 @@
+    protected function configure()
+    {
+        $this
+            ->setName('scheduler:export')
```
Planned for the first point, for the task list, it can be done if needed 🙂

Guikingone replied:
Improved for the first point, second in under progress 🙂

Guikingone commented on src/Symfony/Component/Scheduler/Export/FormatterInterface.php:
```diff
@@ -0,0 +1,33 @@
+interface FormatterInterface
+{
+    public const PRIORITIES = [
+        'ultra_high' => 'ultra_high',
```
In fact, values _will_ be integers, I've started with string but it's not as stable/easy to use as integers 😉

Guikingone replied:
Priorities moved to integers

Guikingone commented on src/Symfony/Component/Scheduler/README.md:
```diff
@@ -0,0 +1,14 @@
+Scheduler Component
+===================
+
+The Scheduler component provides a simple and easy to configure API
```
Yes, my bad :D

Guikingone replied:
Fixed

Guikingone commented on src/Symfony/Component/Scheduler/Task/AbstractTask.php:
```diff
@@ -0,0 +1,174 @@
+     */
+    private $metadata = [];
+
+    public function __construct(string $name, array $metadata = [], array $additionalMetadata = [])
```
I've started the refactoring on this part 🙂

Guikingone replied:
Refactored

Guikingone commented on src/Symfony/Component/Scheduler/Task/TaskPoolInterface.php:
```diff
@@ -0,0 +1,22 @@
+{
+    public function getTasks(): TaskListInterface;
+
+    public function getPriority(): string;
```
TaskPool is removed, the interface should not appear, my bad 😄

Guikingone commented on src/Symfony/Component/Scheduler/Command/RunScheduledTaskCommand.php:
```diff
@@ -0,0 +1,116 @@
+        if (0 === $dueTasks->count()) {
+            $this->style->note('No tasks found');
+
+            return;
```
Fixed

Guikingone commented on src/Symfony/Component/Scheduler/Command/ExportTaskCommand.php:
```diff
@@ -0,0 +1,118 @@
+    /**
+     * @var SymfonyStyle
+     */
+    private $style;
```
Fixed

Guikingone commented:
Hi everyone 👋 

Small update about this PR, I've pushed a "first valid implementation" of the framework configuration (not handled via the container for now) in order to demonstrate how to configure it, some parts are missing (but pushed on secondary repo). 

I've also improved the syntax, some classes and the internal logic of the component, feel free to give feedbacks 🙂

Guikingone commented:
Hi everyone 👋

Small update, I've pushed a first valid configuration along with a first YAML / XML / PHP configuration files in order to validate the current implementation, I'm gonna continue to polish the DX and the configuration in order to finish the "framework" integration.

Small ping to @fabpot (sorry, I really don't want to bother you with stupid questions 😅) in order to know if it worth to continue to work on this PR or if its a better idea to wait for your "cron" component (even if it's planned for 5.2/5.3), the idea behind this ping is really to close this PR if its a bad idea to have 2 "WIP" implementations (I'm pretty sure that yours is probably way more advanced than mine 😄). 

For now, the planning of this PR is to have a "valid" and "ready to be merged" implementation for 5.2 (5.1 is too close IMO) but it all depends on the validity of working on it 🙂 . 

Again, sorry for the ping and have a great evening 🙂  

_PS: I've use this ping because I don't know if there's a "Symfony Core-Team" or "Symfony deciders" one_ 😅

Guikingone commented:
Hi everyone 👋 

Small update, I've pushed a "final" version of the configuration, both YAML & PHP format are supported, the XML one continues to fail (I don't have the reason now). 

The FrameworkBundle integration is almost done (same for the container injection)

Guikingone commented on src/Symfony/Component/Scheduler/Command/ExportTaskCommand.php:
```diff
@@ -0,0 +1,118 @@
+
+        $exportDirectory = sprintf('%s/%s', $exportDirectory, $filename);
+
+        $this->exporter->export($tasks, $exportDirectory, $format);
```
Fixed

Guikingone commented:
Hi everyone 👋 

Small update, I've pushed an improvement on the Scheduler and some additional tests, the FrameworkBundle integration continue to fail (due to XML container configuration, I don't know how to fix it for now 😅 ) but the component is "usable" in a real app (I've tested it in a demo app), I've used Blackfire in order to see if there's some performances issues so far, as far as I can judge, nothing to worry (the CommandRunner still slow) for now.

plozmun commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -2027,6 +2042,133 @@ private function registerNotifierConfiguration(array $config, ContainerBuilder $
+        $container->setParameter('scheduler.timezone', new \DateTimeZone($config['timezone']));
+        $container->setParameter('scheduler.execution_mode', $config['execution_mode']);
+        $container->setParameter('scheduler.export_directory', $config['export_directory']);
+        $allowedModes = ExecutionModeOrchestratorInterface::EXECUTION_MODES;
```
This constant is defined in class and not in the interface

Guikingone replied:
Fixed 🙂

Guikingone commented:
Hi everyone 👋 

Small update on this PR, I've completed the `ExecutionModeOrchestrator` (long name, I agree) logic, for now, it's based on [Linux approach for scheduling process](http://man7.org/linux/man-pages/man7/sched.7.html), the exceptions are: 

- No fork
- No yield
- No autogroups
- "Native" performances

So, now that the "heavy" part is done, I'm focusing on "mixing" everything and connecting everything together, the next part that I'm gonna work on is the "cron" files export, I've already added a "user" option in the Scheduler configuration but that's clearly not enough, so, time to finish it 😄 

Feel free to give feedbacks 🙂

norkunas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/scheduler.xml:
```diff
@@ -20,7 +20,7 @@
         </service>
 
-        <service id="scheduler.cron_warmer" class="Symfony\Component\Scheduler\CronWarmer">
+        <service id="scheduler.cron_warmer" class="CronGenerator">
```
```suggestion
        <service id="scheduler.cron_warmer" class="Symfony\Component\Scheduler\CronGenerator">
```
:slightly_smiling_face:

Guikingone replied:
Fixed 🙂

noniagriconomie commented on src/Symfony/Component/Scheduler/Bridge/Google/Worker.php:
```diff
@@ -0,0 +1,100 @@
+
+        $this->client->run($task->get('configuration'));
+
+        $this->dispatchEvent(new TaskExecutedEvent($task, null));
```
```suggestion
        $this->dispatchEvent(new TaskExecutedEvent($task));
```

Guikingone replied:
Will be removed as Google execute the tasks on its side.

noniagriconomie commented on src/Symfony/Component/Scheduler/Bridge/Google/Worker.php:
```diff
@@ -0,0 +1,100 @@
+            throw new \LogicException('The task state must be defined in order to be executed!');
+        }
+
+        if (\in_array($task->get('state'), [State::PAUSED, State::DISABLED])) {
```
```suggestion
        if (\in_array($task->get('state'), [State::PAUSED, State::DISABLED], true)) {
```

Guikingone replied:
Same as above 🙂

noniagriconomie commented on src/Symfony/Component/Scheduler/Bridge/Google/Worker.php:
```diff
@@ -0,0 +1,100 @@
+     */
+    public function checkTaskState(TaskInterface $task): void
+    {
+        if (State::STATE_UNSPECIFIED === $task->get('state')) {
```
```suggestion
        $taskState = $task->get('state');
        if (State::STATE_UNSPECIFIED === $taskState) {
        // ...
        if (\in_array($taskState, [
```

Guikingone replied:
Same

noniagriconomie commented on src/Symfony/Component/Scheduler/Bridge/Google/Worker.php:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        if (\in_array($task->get('state'), [State::PAUSED, State::DISABLED])) {
+            $this->log(sprintf('The following task "%s" is paused|disabled, consider enable it if it should be executed!', $task->getName()));
```
i would log the task name in the context of the log, and keep the log message identical

Guikingone replied:
Same

noniagriconomie commented on src/Symfony/Component/Scheduler/Bridge/Microsoft/Task/TaskState.php:
```diff
@@ -0,0 +1,23 @@
+    public const DISABLED = 'Disabled';
+    public const ENABLED = 'Enabled';
+    public const FAULTED = 'Faulted';
+}
```
maybe add a private construct? or more those const inside an interface?

Guikingone replied:
Moved to `TaskInterface`

noniagriconomie commented on src/Symfony/Component/Scheduler/Task/AbstractTask.php:
```diff
@@ -0,0 +1,225 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function get(string $key, $default = null)
```
what do you think, for main options, writing direct getters? I saw you use the "state" many times

for example

getState() { return $this->get('state', null); }

noniagriconomie commented on src/Symfony/Component/Scheduler/Task/TaskListInterface.php:
```diff
@@ -0,0 +1,75 @@
+    /**
+     * Add a new task in the list, by default, the name of the task is used as the key.
+     *
+     * @param TaskInterface $task
```
can be removed imo

noniagriconomie commented on src/Symfony/Component/Scheduler/TaskExecutionTracker.php:
```diff
@@ -0,0 +1,50 @@
+/**
+ * @author Guillaume Loulier <contact@guillaumeloulier.fr>
+ */
+final class TaskExecutionTracker
```
```suggestion
final class TaskExecutionWatcher
```

no?

Guikingone replied:
Renamed to `TaskExecutionWatcher`

noniagriconomie commented on src/Symfony/Component/Scheduler/Worker.php:
```diff
@@ -0,0 +1,177 @@
+            throw new \LogicException('The task state must be defined in order to be executed!');
+        }
+
+        if (\in_array($task->get('state'), [State::PAUSED, State::DISABLED])) {
```
same comments as above for the get('state') and the log message + context

noniagriconomie commented on src/Symfony/Component/Scheduler/composer.json:
```diff
@@ -0,0 +1,50 @@
+    "symfony/options-resolver": "^4.4|^5.0",
+    "symfony/process": "^4.4|^5.0",
+    "symfony/serializer": "^4.4|^5.0",
+    "symfony/stopwatch": "^4.4|^5.0"
```
for me the stopwatch should be inside require-dev, and event-dispatcher inside require section, no?

Guikingone replied:
I'm not sure about this one, the EventDispatcher is not required as if you use Scheduler outside of Sf, you can use it without EventDispatcher, the core "logic" does not require the EventDispatcher to run 🤔

Guikingone replied:
The stopwatch has been moved to suggestion, the eventDispatch should stay IMO in the require-dev as the Scheduler can work without it.

noniagriconomie commented on src/Symfony/Component/Scheduler/Worker.php:
```diff
@@ -0,0 +1,177 @@
+    {
+        $this->dispatchEvent(new TaskToExecuteEvent($task));
+        $output = $runner->run($task);
+        $task->set('last_execution', new \DateTimeImmutable());
```
shouldnt this be inside an event subscriber?

Guikingone replied:
Depends on the previous comment

noniagriconomie commented on src/Symfony/Component/Scheduler/Task/State.php:
```diff
@@ -0,0 +1,29 @@
+        self::DISABLED,
+        self::UNSPECIFIED,
+    ];
+}
```
what about adding here private construct

Guikingone replied:
Removed 🙂

Guikingone commented:
Hi everyone 👋 

Small update about this PR, I've improved the configuration of the Scheduler, we now use transports (same logic as the Notifier and Messenger components) to configure the scheduler internal "dispatching", this way, we can improve the bridges and ease the FrameworkBundle integration (using DSN is way easier than configuring every specific connection). 

I've improved the Scheduler logic to use the transports and added a factory that can create every supported transport (including bridges). 

**So, what's next?**

For now, I'm focusing on bringing everything together, improving the framework integration and finishing the GCP bridge (for now, both Azure and AWS are difficult to implement, I may delay the availability of these bridges). 

**So, what about a planning?**

By far, I've planned to be ready for 5.2, however, I can probably push a little bit more and be ready for 5.1.

Guikingone commented:
Hi everyone 👋 

Small update about this PR, I've improved the Scheduler configuration by using transports (like Notifier, Messenger and Mailer), for now, I've implemented both local and null transports, I've created the bridged for Redis and Doctrine, other platforms could be added later (think AWS and Azure). 

I've planned to improved the tasks creation tomorrow (maybe thanks to a similar approach than the transports) and finish the FrameworkBundle integration next week, that's a serious planning but I think I can handle it, I also need to finish the tests and the final benchmarks thanks to Blackfire.

Feel free to give feedbacks 🙂 

@noniagriconomie I've saw your reviews, I need to check what's already fixed and what I've removed 😉

Guikingone commented:
Hi everyone 👋 

Small update about this PR, as promised, I've worked on the task creation (using factories) and finished the transports (the ones which are available "by default", the bridges will follow). 

I've also worked on the FrameworkBundle integration but I face some serious issues with it, it seems that by far, the transports "factories" aren't injected properly using their tags, that's probably an issue when it comes to the `scheduler.xml` file but I've not found anything relevant so I will continue debugging it. 

As far as I know, it seems that the iterable returned thanks to `tagger_iterator` does not contain my transports, strange situation. 

I might finish the "component" aspect next week and focus on the FrameworkBundle and WebProfileBundle integrations in order to finish in 2-3 weeks (depending on the bugs that I'm facing 😄 ). 

Feel free to give feedbacks 🙂

Guikingone commented:
Hi everyone 👋 

Small update about this PR, I've been pretty busy during the last weeks so I wasn't able to work correctly on Scheduler, I've improved the DX and some performances issues but nothing that requires to be pushed by now. 

Here's the planning that I'm trying to target: 

- Symfony 5.2 ready (5.1 is too close)
- Finishing the main transports (GCP, Redis, Doctrine, Azure and Amazone (hard one))
- Finishing the unit tests for the main runners (the callback one need to be improved)
- Solving the issues that I'm facing when it comes to injecting and tagging the transports in FrameworkBundle, by now, it seems that the tag are correctly set but not the services. 

I should be able to get back on track in a few days and finish 2-3 points of the list in 1-2 weeks, by now, I'm targeting a "ready to merge" PR by the end of April). 

Have a great day and don't hesitate to give feedbacks 🙂

Guikingone commented:
Hi everyone 👋 

Small update about this PR, I've FINALLY solve the issue I've faced with FrameworkBundle, the issue was a simple bad usage of `Reference`, nothing too bad 😄 

So, what's next? 

First, I'm on holidays this afternoon, so I should be able to work a little bit harder on this PR (in order to finish it for the end of April), secondly, I'm gonna finish the configuration handling and the services injection/tagging, once done, I should be able to finish the bridges pretty fast (the Google one should be the first to be finished) and focus the finish line on the tests.

Guikingone commented:
Hi everyone 👋 

Hope the week-end goes great for everyone, small update about this PR, the FrameworkBundle integration is now "complete", I need to finish the `SchedulerPass` logic (auto-configuration, etc) but the main integration is done, same things for the main runners/tasks. 

The Google Cloud Platform bridge is almost done, the client is done and I need to finish the DSN handling (options, etc) but everything works as expected. 

The AWS bridge has been removed, I'm not capable of producing it now as I'm clearly not aware of how AWS handle the tasks (it seems that every service has it own "logic", it could become a "post-PR" feature. 

The Azure bridge is in progress even if I'm not sure if it will be available in the final PR, the Scheduler have been moved to `Azure Logic Apps` and the API have changed, I need to investigate on it in order to decide if it worth the work. 

Both Redis & Doctrine transports are under active work, I should be able to finish every "external" transports pretty soon. 

I've improved the core tests and the DX of the PR, here's a sample of what it requires to configure a local task: 

```yaml
framework:
  scheduler:
    timezone: 'Europe/Paris'
    transports:
      local:
        dsn: 'local://root'
    schedulers:
      foo:
        transport: 'local'
    tasks:
      app.foo:
        scheduler: 'foo'
        type: 'shell'
        expression: '* * * * *'
        options:
          command: 'echo Symfony!'
```

That's all for today, hope the work that I'm putting on it continue to make this PR relevant 🙂

Guikingone commented:
Hi everyone 👋 

Hope the week-end goes well for everyone 🙂 

Small update about this PR, I'm happy to announce that the Google client is "officially" done, I need to finish some tests and error handling but the core logic is in place, the DX must be improved (mainly for the DSN handling). 

Regarding the Redis & Doctrine transports, I progress slowly, the Doctrine one need to be improved and the Redis one must be finished. 

So, what about the April deadline?
For now, I keep it in mind and try to focus in order to finish it for it, I will probably need 1-2 additional weeks to finish this PR (as I'm trying to add round-robin & failover transports) and the related tests, I also consider adding the ["preloading tags"](https://symfony.com/blog/new-in-symfony-5-1-configurable-php-preloading). 

I'm also considering dropping the AWS transport for now, the API has recently changed and it seems that every services define its own tasks, it could be hard to track the changes and I'm not convinced about the use cases for now.

That's all for today, feel free to give feedbacks 🙂

94noni commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -48,6 +48,7 @@
         "symfony/messenger": "^4.4|^5.0",
         "symfony/mime": "^4.4|^5.0",
         "symfony/process": "^4.4|^5.0",
+        "symfony/scheduler": "^4.4|^5.0",
```
shouldn't be 5.1 ?

derrabus replied:
```suggestion
        "symfony/scheduler": "^5.2",
```

There is no version 4.4 of that package.

94noni commented on src/Symfony/Component/Scheduler/Bridge/Doctrine/Transport/Connection.php:
```diff
@@ -0,0 +1,183 @@
+    private function createQueryBuilder(): QueryBuilder
+    {
+        return $this->driverConnection->createQueryBuilder()
+            ->select('m.*')
```
what is the m? maybe use task or schedule?

Guikingone replied:
Fixed

94noni commented on src/Symfony/Component/Scheduler/Bridge/Microsoft/Transport/AzureTransportFactory.php:
```diff
@@ -0,0 +1,39 @@
+     */
+    public function support(string $dsn, array $options = []): bool
+    {
+        return 0 == strpos($dsn, 'azure://');
```
maybe use strict comparison?

94noni commented on src/Symfony/Component/Scheduler/Bridge/Redis/Transport/Connection.php:
```diff
@@ -0,0 +1,97 @@
+            throw new LogicException('The redis transport requires php-redis 4.3.0 or higher.');
+        }
+
+        $this->connection = $redis ?: new \Redis();
```
```suggestion
        $this->connection = $redis ?? new \Redis();
```

no?

Guikingone replied:
Fixed

94noni commented on src/Symfony/Component/Scheduler/Bridge/Redis/Transport/Connection.php:
```diff
@@ -0,0 +1,97 @@
+final class Connection
+{
+    private const DEFAULT_OPTIONS = [
+        'address' => '127.0.0.1',
```
why not host as key?

Guikingone replied:
Of course

94noni commented on src/Symfony/Component/Scheduler/Command/ExportTaskCommand.php:
```diff
@@ -0,0 +1,117 @@
+
+        $this->exporter->export($tasks, $exportDirectory, $format);
+        $this->io->success(sprintf('Exported "%d" tasks to "%s"', \count($tasks), sprintf('%s.%s', $exportDirectory, $format)));
+    }
```
you must return a bool here :)

Guikingone replied:
Fixed

94noni commented on src/Symfony/Component/Scheduler/Command/RebootSchedulerCommand.php:
```diff
@@ -0,0 +1,82 @@
+    {
+        $name = $input->getArgument('name');
+
+        $scheduler = $this->container->get('scheduler.hub.'.$name);
```
maybe have a scheduler registry instead of injecting the container?

Guikingone replied:
I completely agree with this idea, I've implemented this idea, huge thanks 🙂

94noni commented on src/Symfony/Component/Scheduler/Command/RunScheduledTaskCommand.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * {@inheritdoc}
+     */
+    protected function interact(InputInterface $input, OutputInterface $output): void
```
is interact the right method here?
seems not interacting with enduser

94noni commented on src/Symfony/Component/Scheduler/EventListener/TaskSubscriber.php:
```diff
@@ -0,0 +1,90 @@
+ */
+final class TaskSubscriber implements EventSubscriberInterface
+{
+    private $schedulers;
```
maybe add php doc to know the type behinds the iterable

94noni commented on src/Symfony/Component/Scheduler/EventListener/TaskSubscriber.php:
```diff
@@ -0,0 +1,90 @@
+        }
+
+        foreach ($tasks as $task) {
+            $this->worker->execute($task);
```
this seems dangerous
maybe add a role or a token to "securise" this GET action?

also, if a task takes longer than the web timeout what happened?

Guikingone replied:
I'm not sure about how to do it right here, IMO, the security of this route should be a charge of the developer (maybe via JWT, OAuth or other?), I also can ask for a specific header but I'm not sure 🙁 

I'm gonna take a look about the timeout case 🙂

94noni commented on src/Symfony/Component/Scheduler/Exception/AlreadyScheduledTaskException.php:
```diff
@@ -0,0 +1,19 @@
+/**
+ * @author Guillaume Loulier <contact@guillaumeloulier.fr>
+ */
+final class AlreadyScheduledTaskException extends \RuntimeException
```
maybe you should/could create a base class ex of the component
so that an external dev can catches all ex from this component easily

94noni commented on src/Symfony/Component/Scheduler/WorkerInterface.php:
```diff
@@ -0,0 +1,33 @@
+    /**
+     * Execute the given task, if the task cannot be executed, the worker should exit.
+     *
+     * An exception can be throw during the execution of the task, if so, it SHOULD be handled.
```
maybe defined a WorkerExceptionInterface no?

94noni commented on src/Symfony/Component/Scheduler/README.md:
```diff
@@ -0,0 +1,19 @@
+Scheduler Component
```
maybe add also a default/basic changelog?

Guikingone replied:
As I'm not sure if this PR will be accepted, I'm not sure about adding a changelog as this code is not publicly available, at least for now 🤔

Guikingone commented:
Hi @94noni 👋 

Thanks for the reviews, I take a look at it tomorrow morning 🙂

norkunas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1569,4 +1571,96 @@ private function addNotifierSection(ArrayNodeDefinition $rootNode)
+                            ->defaultValue('/_tasks')
+                        ->end()
+                        ->scalarNode('timezone')
+                            ->defaultValue('Europe/Paris')
```
Maybe `utc` would fit better?

Guikingone replied:
Thanks for the spot, fixed on the latest commit 🙂

Guikingone commented:
Hi everyone 👋 

It's been a long time since the last update on this PR, thanks to the lockdown, I'm managed to work on core refactoring and improvements, for now, here's what missing: 

- Doctrine | Redis transports
- Azure transport abandonned
- DX improvements (mostly about alias registration, etc)
- Some tests on the core logic (mostly execution commands and tasks handling)

I've fixed @norkunas review and I need to finish some improvements highlighted by @94noni (thanks both of you for the reviews 😊 ). 

Before going further, I think it's time to ping @fabpot and/or @nicolas-grekas about this PR, why? 

First, I've saw that https://github.com/symfony/symfony/pull/36778 have been opened, If I'm right, this PR tries to removed all the XML configuration of Symfony (a good point IMO), the question is, as the current PR use the XML configuration (for both container and services definition), is it a good idea to continue to work on it and maintain the XML configuration (even if the core changes are planned for Sf 6)? 

Second, as we know that Fabpot is working on a Cron component and will probably finish it way faster than this PR, is it a good idea to keep two "components" under progress? 

This current PR is probably miles away for the quality required by Symfony and I know that the DX need to be drastically improved, so, I'm not sure if it worth the work to continue working on this PR and if it's not a better idea to wait for the Fabpot Cron component and maybe contribute on it when it's released? 

Thanks for the feedback and have a great day 🙂

Guikingone commented:
Hi everyone 👋 

Small update about this PR, here the latest features implemented: 

- Commands for both cron generation and failed tasks listing
- Cron registry added (for storing every cron generated for the schedulers)
- Tests added for main features (commands one improved)

Here's what's missing: 

- Doctrine / Redis transport
- Fix on cron expression validation
- Security on TaskListener 

Thanks again for the feedback and have a great day 🙂

Guikingone committed:
refactor(bridge): progress on Kubernetes connection

Guikingone committed:
feat(core): SchedulerAwareInterface added & configured

Guikingone commented:
Hi everyone 👋 

New week, new update, I've mainly worked on transport and DX, here's the "changelog" for this week-end: 

- Kernel & classes as Scheduler "entrypoints"

`Scheduler` defines a `SchedulerAwareInterface` which allow virtually any class to be a entrypoint to define new tasks: 

```php
<?php

/*
 * This file is part of the Symfony package.
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Symfony\Component\Scheduler;

/**
 * @author Guillaume Loulier <contact@guillaumeloulier.fr>
 */
interface SchedulerAwareInterface
{
    public function schedule(SchedulerRegistryInterface $registry): void;
}
```

This interface is auto configured and every "class" that implement in will receive the `SchedulerRegistry` which contain every declared scheduler. 

Useful for commands, message handler, kernel and so on 🙂 

- New transports: 

Kubernetes started, Redis & Doctrine under active work, `FileTransport` started (think of it as the `MockFileSessionStorage` for Scheduler) 🙂 

Improvements on core commands (under active tests). 

Thanks again for the feedback and have a great day 🙂

Guikingone committed:
tests(core): tests improved for NullTaskRunner

Guikingone commented:
Hi everyone 👋 

Second update of the day: 

- A `--dry-run` option has been added to `GenerateCronCommand` and `RebootSchedulerCommand`
- Tests improved for `ConsumeTasksCommand`

Thanks again for the feedback and have a great day 🙂

Guikingone committed:
fix(bundle): UnusedTag fixed

Guikingone committed:
refactor(bridge): Doctrine under work

Guikingone committed:
feat(bridge): Doctrine call under work

Guikingone committed:
style(bridge): Doctrine connection fixed

Guikingone committed:
feat(bundle): FrameworkExtension improved with bridges services injection

Guikingone commented:
Hi everyone 👋 

Small update for today, I've almost finished the Doctrine transport, I need to implement the sorting part of every task stored in the transport (same logic as the `local` transport) but the core logic is done and tested. 

I've also implemented the bridge transport factories on the FrameworkExtension thanks to a specific configuration file. 

Thanks again for the feedback and have a great day 🙂

Guikingone commented:
Hi everyone 👋 

New day, new update, first, I can officially announce that I'm close to a "ready to review" PR, the `Doctrine` transport is done and tested, the `Redis` one is close to a complete implementation (I need to focus on the core methods) and the tests should be done pretty soon. 

Other news: 

- The `Kubernetes` transport is under work (I need to finish the DSN handling)
- The tasks have been refactored when it comes to structure and usage (a huge thank to @kbond for his work on the bundle, I've managed to implement a similar approach (not completely the same as I have a different structure))
- The `file` transport will be removed, I don't see any use case (except `test` env but I already have implemented a TraceableScheduler/Transport so there's no need for it). 

I also have some ideas when it comes to "recent" components, I've planned to work on this "extra idea" in parallel of the transports, I should be able to give more info very soon. 

Again, thanks for the feedback and support and have a great day 🙂

Guikingone commented:
Hi everyone 👋 

Small update, the Redis transport is now done and tested, I continue to work on the Kubernetes one and implements core details.

I've also implemented a `WorkerSubscriberInterface` which allow to register subscriber for specific/every worker (as we have a worker per scheduler). 

Next goals: 

- Add a Notifier support (when it comes to notifying the success/failure of a task)
- Add a Mailer support (when it comes to notifying the success/failure of a task)
- Finishing the Kubernetes transport
- Adding security on the TaskSubscriber

Again, thanks for the feedback and support and have a great day 🙂

Guikingone commented:
@94noni I've requested a small review as the PR as evolved and as I've implemented specific interfaces for exceptions 🙂

Guikingone committed:
fix(bundle): fix on UnusedTagPass

Guikingone commented:
Hi everyone 👋 

New day, new update, pretty big API changement here, first, the concept of multiples workers has been removed, there's now only one worker for the whole "component", I'm thinking about playing with pnctl in order to scale the worker as the need grows but nothing "physical" with a `WorkerRegistry`. 

Second point, I've removed the `TraceableWorker` as there's no interest for now. 

Third point, I've was this morning that Ryan [has merged](https://symfony.com/blog/new-in-symfony-5-1-automatic-schema-updates) an important feature in Doctrine bridge for external transport (especially the Doctrine ones), I'm pretty happy to announce that this PR already support it thanks to the Doctrine bridge for Scheduler. 

What else ? Hum ...

Important feature! 

I've worked a lot with Messenger recently on side projects and I've been impressed with the stamp concept (which are not "usable" as it in this PR), so I've worker on a similar concept called `Bag`, so, what's the point here? 

Like @kbond has done it in his bundle thanks to "extensions" (for http call, emails, etc), I've managed to implement similar behaviour via bags, here's an example for notifications (thanks to Fabpot Notifier): 

```php
$preExecutionNotification = new Notification(...);

$scheduler = new Scheduler(...);
$scheduler->schedule(new NullTask('foo', new NotifierBag([
     $preExecutionNotification // will be sent before the task
])));
```

Here's the full API for the `NotifierBag`: 

```php
x<?php

/*
 * This file is part of the Symfony package.
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Symfony\Component\Scheduler\Bag;

use Symfony\Component\Notifier\Notification\Notification;

/**
 * @author Guillaume Loulier <contact@guillaumeloulier.fr>
 */
final class NotifierBag implements BagInterface
{
    private $notifications;

    /**
     * @param array|Notification[] $beforeNotifications
     * @param array|Notification[] $afterNotifications
     * @param array|Notification[] $failureNotifications
     */
    public function __construct(array $beforeNotifications = [], array $afterNotifications = [], array $failureNotifications = [])
    {
        $this->notifications['before'] = $beforeNotifications;
        $this->notifications['after'] = $afterNotifications;
        $this->notifications['failure'] = $failureNotifications;
    }

    public function getNotifications(): array
    {
        return $this->notifications;
    }

    /**
     * {@inheritdoc}
     */
    public function getName(): string
    {
        return 'notifier';
    }
}
```
A similar bag has been created for `Mailer` and I'm working on a specific one for `HttpClient` (what about `Mercure` from Dunglas? 😉 ), every bag has its own subscriber which listen specific event and call the dedicated component to use it, so the API is pretty easy to extends 🙂 .

Guikingone commented:
Hi everyone 👋 

New day, new update (small one for today), I've mainly worked on the core feature like Bag's, subscribers and tasks: 

- Improvements on bags: HttpClientBag removed, Mercure and Messenger tested
- Fix on task lifecycle subscribers
- HttpTask introduced (use the `HttpClient` to make request)

I've also worked on the feature asked by @monteiro on January 19th: 

- One run tasks (his example about user engagement is a pretty good one)

For now, I'm not sure about the implementation, here's what I think that could be a good idea: 

- Defining a `SingleRunTask` which accept a `Task` (which is the one that we will run) and an expression (defined as a `DateInterval` maybe?)

IMO, that's the easiest way of doing it and we can easily target this task in order to remove it from the Scheduler once executed, any idea? 

Thanks for the feedback and support and have a great day 🙂

Guikingone commented:
Hi everyone 👋 

New day, new update, small one for today, I've refactored the `Redis` transport in order to use the Symfony Serializer and a custom normalizer/denormalizer (which can handle virtually any task type). 

I need to refactor the core API in order to make the tests goes green but that should be fast, I've also refactored the `Kubernetes` transport to use a custom normalizer/denormalizer (the `Google` transport will probably receive the same approach). 

That's all for today, thanks for the feedback and support and have a great day 🙂

Guikingone commented:
Hi everyone 👋 

New day, new update (again, I know 😄 ), after yesterday refactoring on `Redis` transport, I've moved the work on the `Kubernetes` transport, for now, the transport is ready, I need to test it completely and ensure that everything goes well. 

Second point, I'm gonna work on an update of the `Doctrine` transport, for now, the actual implementation is "great" but not as clean as I wanted to make it, I'm gonna use the `Serializer` and a similar approach as the one used by Redis. 

Third point, I'm also gonna work on the last tests to be implemented and some improvements, I'm not gonna define a "ready to review" date as I probably not gonna work on this PR for 2-3 days this week (due to several reasons) but it should be ready by the end June for sure (I hope 😄 ).  

That's all for today, thanks for the feedback and support and have a great day 🙂

Guikingone commented:
Hi everyone 👋 

Small update about this PR, I've refactored the `Redis` transport in order to implement the orchestrator (the layer that can sort tasks when inserted), the Doctrine transport will be refactored on the same way. 

I've also worked on the `Nomad` and `Kubernetes` transport and started the `Dkron` one. 

I've also talked with Nicolas Grekas about the future of this PR, for now, nothing is really fixed, it's open to debate and I don't know if it will be part of Symfony.
I'm not sure if it worth to create a separate package (as other framework already implement this type of component) with a simple bundle that wrap default layer within Symfony, so for now, this PR will continue to be improved and I've planned to make it "ready to review/debate" for June, please note that it will probably not perfect when this time will arrives, I'm fully open to reviews and good/bad feedback on this PR if you have time 🙂 

That's all for today, thanks for the feedback and support and have a great day 🙂

noniagriconomie commented:
hi @Guikingone I remembered this PR this morning, and would like to know its state in the release process of 5.2

- would it be inside as experimental?
  - maybe like the notifier, it can be merged without all the bridges for now
- will it be closed in favor of other codebase?

many thanks, cheers :)

Guikingone commented:
Hi @noniagriconomie 👋 

Actually, I don't have informations about a release in `5.2`, experimental approach or other codebase for now, I've asked @nicolas-grekas about it but at the moment of the question (middle of June if I'm right), nothing was planned and the core team needed time to inspect the code and see what to do with the PR. 

So for now, I don't know what to say about this PR, the core logic is implemented and most of the bridges too (maybe the Doctrine one need to be improved but it could easily be done in time for `5.2`) so on my side, it's "ready" for reviews and improvements but I cannot gave you more informations as I don't have much on my side 😅 .

fabpot commented:
@Guikingone I will review this PR soon(tm). Can you remove the Draft status and the WIP in the description if it's ready?

fabpot commented:
Also, can you rebase on current master?

noniagriconomie commented:
@Guikingone fair enough :)
I will be glad to make another round of reviews when the PR is ready!
(personal ping @94noni)

Guikingone commented:
PR rebased and ready for reviews

derrabus commented on composer.json:
```diff
@@ -18,6 +18,7 @@
     "require": {
         "php": ">=7.2.5",
         "ext-xml": "*",
+        "dragonmantank/cron-expression": "^2.3",
```
On the root composer.json, this dependency should probably be moved to `require-dev`. Also, there's a version 3 of that library available. Can we upgrade?

Guikingone replied:
Moved to `3.0`

derrabus commented on src/Symfony/Component/Scheduler/Bag/BagRegistry.php:
```diff
@@ -0,0 +1,90 @@
+ */
+final class BagRegistry implements BagRegistryInterface
+{
+    private $bags = [];
```
```suggestion
    /**
     * @var BagInterface[]
     */
    private $bags = [];
```

derrabus commented on src/Symfony/Component/Scheduler/Bag/MailerBag.php:
```diff
@@ -0,0 +1,50 @@
+    private $mails;
+
+    /**
+     * @param array|RawMessage[] $beforeMails
+     * @param array|RawMessage[] $afterMails
+     * @param array|RawMessage[] $failureMails
```
```suggestion
     * @param RawMessage[] $beforeMails
     * @param RawMessage[] $afterMails
     * @param RawMessage[] $failureMails
```

Guikingone replied:
Improved

94noni commented on src/Symfony/Component/Scheduler/Export/ExporterInterface.php:
```diff
@@ -0,0 +1,25 @@
+ */
+interface ExporterInterface
+{
+    public function exportSingle(TaskInterface $task, string $filePath, string $format): void;
```
i  would do the opposite, export() => one and exportMultiple|Many() => multiple|many

Guikingone replied:
Renamed

94noni commented on src/Symfony/Component/Scheduler/EventListener/TaskSubscriber.php:
```diff
@@ -0,0 +1,95 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Scheduler\EventListener;
```
why not inside a `Scheduler/EventSubscriber` folder as its a Subscriber?

94noni commented on src/Symfony/Component/Scheduler/EventListener/StopWorkerOnTimeLimitSubscriber.php:
```diff
@@ -0,0 +1,62 @@
+
+    public function onWorkerStarted(WorkerStartedEvent $event): void
+    {
+
```
did you forgot to call the private log function on both methods?

94noni reviewed (commented):
First small round :)

Guikingone commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/scheduler.xml:
```diff
@@ -0,0 +1,140 @@
+<?xml version="1.0" ?>
```
Fixed, nice one

Guikingone commented on src/Symfony/Component/Scheduler/Bag/BagRegistry.php:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function count()
```
Fixed

Guikingone commented on src/Symfony/Component/Scheduler/composer.json:
```diff
@@ -0,0 +1,64 @@
+  "name": "symfony/scheduler",
+  "type": "library",
+  "description": "Symfony Scheduler Component",
+  "keywords": ["schedule", "scheduler", "tasks", "cron"],
```
Added

Guikingone commented on src/Symfony/Component/Scheduler/Transport/TransportInterface.php:
```diff
@@ -0,0 +1,39 @@
+
+    public function resume(string $taskName): void;
+
+    public function empty(): void;
```
Renamed

Guikingone commented on src/Symfony/Component/Scheduler/Runner/CommandTaskRunner.php:
```diff
@@ -0,0 +1,108 @@
+     */
+    public function run(TaskInterface $task): Output
+    {
+        if (!$task instanceof CommandTask) {
```
Yep, fixed

Guikingone commented on src/Symfony/Component/Scheduler/Runner/MessengerTaskRunner.php:
```diff
@@ -0,0 +1,56 @@
+            $this->bus->dispatch($task->get('message'));
+
+            return Output::forSuccess($task, 0, null);
+        } catch (\Throwable | \Exception $exception) {
```
Removed `\Exception`

Guikingone committed:
style(PR): fabbot fixs

Guikingone commented on src/Symfony/Component/Scheduler/Worker/WorkerInterface.php:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Execute the given task, if the task cannot be executed, the worker should exit.
+     *
+     * An exception can be throw during the execution of the task, if so, it SHOULD be handled.
```
Fixed

Guikingone committed:
style(PR): fabbot fixs

94noni reviewed (commented):
Second round :)

Guikingone committed:
refactor(Scheduler): reviews

Guikingone commented on src/Symfony/Component/Scheduler/Event/WorkerStartedEvent.php:
```diff
@@ -0,0 +1,40 @@
+        return $this->worker;
+    }
+
+    public function isWorkerIdle(): bool
```
Renamed

Guikingone commented on src/Symfony/Component/Scheduler/Worker/Worker.php:
```diff
@@ -0,0 +1,217 @@
+                }
+            }
+
+            throw new UndefinedRunnerException(sprintf('No runner found for the given task "%s"', $task->getName()));
```
Reworded

Guikingone commented on src/Symfony/Component/Scheduler/Bridge/Redis/Transport/Connection.php:
```diff
@@ -0,0 +1,253 @@
+        'host' => '127.0.0.1',
+        'port'    => 6379,
+        'timeout' => 30,
+        'dbindex' => 'scheduler_tasks',
```
Renamed

Guikingone committed:
refactor(Scheduler): fixs

Guikingone commented on src/Symfony/Component/Scheduler/Command/GenerateCronCommand.php:
```diff
@@ -0,0 +1,112 @@
+    private $generator;
+    private $registry;
+
+    protected static $defaultName = 'scheduler:generate';
```
Renamed

Guikingone committed:
refactor(Scheduler) reviews

Guikingone commented on src/Symfony/Component/Scheduler/Messenger/TaskMessageHandler.php:
```diff
@@ -0,0 +1,47 @@
+        }
+
+        while ($this->worker->isRunning()) {
+            sleep(0.5);
```
Improved

94noni commented on src/Symfony/Component/Scheduler/Messenger/TaskMessage.php:
```diff
@@ -0,0 +1,46 @@
+     */
+    private $workerTimeout;
+
+    public function __construct(TaskInterface $task, $workerTimeout = 0.5)
```
watch out, `sleep()` waits an int, so need here `int $workerTimeout = 1`
or uses `usleep()` :)

Guikingone committed:
refactor(Scheduler): improvements|naming improved

Guikingone commented on src/Symfony/Component/Scheduler/Event/TaskToExecuteEvent.php:
```diff
@@ -0,0 +1,33 @@
+/**
+ * @author Guillaume Loulier <contact@guillaumeloulier.fr>
+ */
+final class TaskToExecuteEvent extends Event
```
Renamed

Guikingone commented on src/Symfony/Component/Scheduler/Task/Output.php:
```diff
@@ -0,0 +1,80 @@
+    private $task;
+    private $type;
+
+    public function __construct(TaskInterface $task, int $exitCode, ?string $output, string $type)
```
Improved with default value

Guikingone commented on src/Symfony/Component/Scheduler/Cron/CronGenerator.php:
```diff
@@ -0,0 +1,61 @@
+        $this->logger = $logger;
+    }
+
+    public function generate(string $filename = 'app', string $path = self::DEFAULT_PATH): void
```
Renamed

fabpot commented:
Closing this PR for now. @Guikingone and I will work together to submit a new PR with some substantial changes.
